Kernel/Devices/VMWareSVGADevice.cpp

```diff
@@ -0,0 +1,283 @@
+        if (!copy_from_user(&resolution, user_resolution))
+            return -EFAULT;
+        set_mode(resolution.width, resolution.height, 32);
+        resolution.width = m_framebuffer_width;
+        resolution.height = m_framebuffer_height;
+        resolution.pitch = m_framebuffer_pitch;
+        if (!copy_to_user(user_resolution, &resolution))
+            return -EFAULT;
+        return 0;
+    }
+
+    default: {
+        return -EINVAL;
+    }
+    }
+}
+
+KResultOr<Region*> VMWareSVGADevice::mmap(Process& process, FileDescription&, const Range& range, u64 offset, int prot, bool shared)
+{
+    REQUIRE_PROMISE(video);
+    if (!shared)
+        return ENODEV;
+    if (offset != 0)
+        return ENXIO;
+    if (range.size() != page_round_up(m_framebuffer_size))
+        return EOVERFLOW;
+
+    auto vmobject = AnonymousVMObject::create_for_physical_range(m_framebuffer_address_physical, m_framebuffer_size);
+    if (!vmobject)
+        return ENOMEM;
+    return process.space().allocate_region_with_vmobject(
+        range,
+        vmobject.release_nonnull(),
+        0,
+        "VMWare SVGA Framebuffer",
+        prot,
+        shared);
+}
+
+String VMWareSVGADevice::device_name() const
+{
+    return String::formatted("fb{}", minor());
+}
+
+void VMWareSVGADevice::handle_irq(const RegisterState&)
+{
+    // Read and clear the interrupts
+    u16 irq_port = m_io_base + static_cast<u32>(Ports::IRQStatus);
+    u32 irqs = IO::in32(irq_port);
+    IO::out32(irq_port, irqs);
+
+    if(!irqs) {
+        dbgln("VMWareSVGA: received interrupt but no bit in IRQStatus is set - spurious or not for this device");
+        return;
+    }
+    // TODO: IRQs should be a bitmask here, but there was no entry for vsync, and all the interrupts are masked?
+    if (m_mode_set)
+        force_vblank();
+}
+
+u32* VMWareSVGADevice::reserve_fifo(size_t size)
+{
+    m_fifo_bounce_buffer_used = true;
+    m_fifo_bounce_buffer.ensure_capacity(size);
+    m_fifo_bounce_buffer.resize(size, true);
+    return m_fifo_bounce_buffer.data();
+}
+
+void VMWareSVGADevice::commit_fifo(size_t size)
+{
+    u32 next_cmd_in_fifo = read_fifo(FifoRegister::NextCMD); // Offset to the next command in the fifo
+    u32 min_fifo = read_fifo(FifoRegister::Min);             // Offset to the beginning of usable fifo memory
+    u32 max_fifo = read_fifo(FifoRegister::Max);             // Offset to the end of usable fifo memory
+
+    // TODO: Support FifoCapability::Reserve
+    // It would be faster, but Qemu doesnt support it (or at least doesnt report support for it), so testing it is a bit on the hard side
+    for (size_t i = 0; i < size; i++) {
+        write_fifo(next_cmd_in_fifo / sizeof(u32), m_fifo_bounce_buffer[i]);
+        next_cmd_in_fifo += sizeof(u32);
+        if (next_cmd_in_fifo == max_fifo) {
+            next_cmd_in_fifo = min_fifo;
+        }
+        write_fifo(FifoRegister::NextCMD, next_cmd_in_fifo);
+    }
+    m_fifo_bounce_buffer_used = false;
```

Comment:
I don't think you need to track `m_fifo_bounce_buffer_used` separately. `m_fifo_bound_buffer.size() > 0` implies this.